MathPortal/MathPortal/Source/Custom/Custom Views/SolutionView.swift

```diff
@@ -0,0 +1,79 @@
+//
+//  SolutionView.swift
+//  MathPortal
+//
+//  Created by Petra Čačkov on 03/10/2019.
+//  Copyright © 2019 Petra Čačkov. All rights reserved.
+//
+
+import UIKit
+
+class SolutionView: UIView {
+    
+    var solution: TaskSolution?
+    private var equationViewsHeight: CGFloat = 0
+    private var ownerInfoViewHeight: CGFloat = 60
+    // NOTE: 2 is just a placeholder since cell already has width
+    override var intrinsicContentSize: CGSize {
+        return CGSize(width: 2, height: equationViewsHeight + ownerInfoViewHeight)
+    }
+    
+    func setupView(solution: TaskSolution) {
+        self.solution = solution
+        removeSubviews()
+        addViews()
+        setupFrame()
+        fetchSolutionOwner(solution: solution)
+        self.invalidateIntrinsicContentSize()
+    }
+    
+    private func removeSubviews() {
+        for view in subviews {
+            view.removeFromSuperview()
+        }
+    }
+    
+    private func fetchSolutionOwner(solution: TaskSolution) {
+        solution.fetchSolutionOwner { (owner, error) in
+            guard self.solution?.objectId == solution.objectId else { return }
+            if let owner = owner {
+                self.addProfileInfo(owner)
+            } else {
+                print(error?.localizedDescription)
+            }
+        }
+    }
+    
+    private func addViews() {
+        var yPosition: CGFloat = 10
+        let xPosition: CGFloat = 10
+        solution?.equations?.forEach { equation in
+            guard let equationView = equation.expression.generateView().view else { return }
+            equationView.frame.origin = CGPoint(x: xPosition, y: yPosition)
+            yPosition += equationView.bounds.height + 10
+            self.addSubview(equationView)
+        }
+        equationViewsHeight = yPosition
+    }
+    
+    private func addProfileInfo(_ owner: User) {
+        
+        let profileImageView: UIImageView = UIImageView()
+        profileImageView.frame = CGRect(x: 0, y: equationViewsHeight, width: ownerInfoViewHeight, height: ownerInfoViewHeight)
+        profileImageView.image = owner.profileImage
+        
+        let usernameLabel: UILabel = UILabel()
+        usernameLabel.text = owner.username
+        usernameLabel.textColor = .black
+        usernameLabel.sizeToFit()
+        usernameLabel.center = CGPoint(x: profileImageView.frame.maxX + 20, y: profileImageView.frame.origin.y + profileImageView.bounds.height/2)
+        
+        self.addSubview(usernameLabel)
```

Comment:
I think this has a possibility to have multiple instances on a view at the same time. Why don't you use lazy properties in this case to avoid it?

Reply:
i don't get it 😔

Reply:
You have a call to setupViews which removes all subviews. Then you do an asynchronous call to get owner which then creates a new view and adds it as a subview. If user would quickly call setupView twice (scrolls table view where this view is on every cell) you may have calls in order:

removeSubviews
removeSubviews
addProfileInfo
addProfileInfo

In this case you would again have 2 labels added at the same time (also other views). You could for instance do:

```
lazy private var usernameLabel: UILabel = {
    let label = UILabel()
    self.addSubview(label)
    return label
}()
```

This would then prevent having multiple labels.

Note: The situation I have described can actually only happen if the same solution is loaded twice. This would probably make the issue invisible but it is still there...